### PR TITLE
WIP: Windows support  -- depends on Meson build system (#3160)

### PR DESCRIPTION
**Pull request by Ericson2314**
Credits to @volth and @p01arst0rm for doing the real work. I just fixed some merge conflicts, fixed linux regressions, and unified the build systems. [Will delete @volth's temporary build system once I can get things building with https://github.com/NixOS/nixpkgs/pull/72366.]

The history is....wild, but if someone want's that fixed, I don't think it's worth fixing until this is closer to being ready.

### CODE REVIEW

**Comment by volth** [comment]
wow

**Comment by volth** [comment]
My idea for the merge was to slowly converge nix-linux and nix-windows, first by small PRs https://github.com/NixOS/nix/issues/2591 https://github.com/NixOS/nix/issues/2603 https://github.com/NixOS/nix/issues/2634 then replacing `Path` with a platform-independent version(probably from `boost`), and then, voila, the diff is trivial.
But that got stuck.

**Comment by Ericson2314** [comment]
@volth that sounds great. I'd add the meson change as another such prereq, but I think we should have this big monster open too as we converge so we know what we are converging and trakc progress.

**Comment by Ericson2314** [comment]
> Can you summarize what this does?

That's for @volth. I hardly know!

> A quick comment after having a glance at the diff: I think for the long-term maintainability of Nix it's important not to have the code littered with `#ifndef _WIN32` / `_MSC_VER`. So we should abstract over these incompatibilities in some way.

Agreed! That relates to what @volth said too. My plan is lets keep this open as a proof of concept, but only merge once everything is better abstracted and the diff small due to newer since-merged PRs. 

Also, I think the abstractions we come up alone with will be half the benefit of this work. It should inform the architectural questions of Nix we had on the flakes RFC for example.

[Also BTW check out http://ckaestne.github.io/TypeChef/, maybe someday we can do that nixpkgs wide and CPP will be a *lot* more maintainable! That's another long-term goal of mine.]

**Comment by edolstra** [comment]
BTW, on the subject of abstracting over (store) paths, the Rust branch has a `StorePath` type which is basically a tuple `(hash, name)` and is enforced to be correct by its constructor (so you can only pass "correct" store paths). It also doesn't store the `/nix/store` prefix.

**Comment by volth** [comment]
Yes, c:\nix\store for a while

Abstraction is a good idea, someone may want it on d:\nix\store or
\\server\nix\store

**Comment by p01arst0rm** [comment]
> Yes, c:\nix\store for a while Abstraction is a good idea, someone may want it on d:\nix\store or \\server\nix\store
> […](#)
> On 10/31/19, Eelco Dolstra ***@***.***> wrote: BTW, on the subject of abstracting over (store) paths, the Rust branch has a `StorePath` type which is basically a tuple `(hash, name)` and is enforced to be correct by its constructor (so you can only pass "correct" store paths). It also doesn't store the `/nix/store` prefix. -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#3185 (comment)](https://github.com/NixOS/nix/pull/3185#issuecomment-548449035)

surely this would be easy to do as  a cxxflag passed into the build(s)? iirc thats how its done for a bunch of stuff. that way you can determine the default location in the config options

**Comment by volth** [comment]
> > Yes, c:\nix\store for a while Abstraction is a good idea, someone may want it on d:\nix\store or \server\nix\store
> > […](#)
> > On 10/31/19, Eelco Dolstra _**@**_.***> wrote: BTW, on the subject of abstracting over (store) paths, the Rust branch has a `StorePath` type which is basically a tuple `(hash, name)` and is enforced to be correct by its constructor (so you can only pass "correct" store paths). It also doesn't store the `/nix/store` prefix. -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#3185 (comment)](https://github.com/NixOS/nix/pull/3185#issuecomment-548449035)
> 
> surely this would be easy to do as a cxxflag passed into the build(s)? iirc thats how its done for a bunch of stuff. that way you can determine the default location in the config options

[It is configured in `Makefile`](https://github.com/NixOS/nix/pull/3185/files#diff-a2f2a3db29e468ef4d80b60ae2ee1b4eR5-R10), not hardcoded in C++
But that `Makefile` was made only for developing purposes, it will be replaced with meson, right?
When Nix build using Nix there is a configurable option passed down:

https://github.com/NixOS/nixpkgs/blob/fc2592b2d6348f73bfc7150f71c3df7cec0bdcd2/pkgs/tools/package-management/nix/default.nix#L2

**Comment by p01arst0rm** [comment]
> > > Yes, c:\nix\store for a while Abstraction is a good idea, someone may want it on d:\nix\store or \server\nix\store
> > > […](#)
> > > On 10/31/19, Eelco Dolstra _**@**_.***> wrote: BTW, on the subject of abstracting over (store) paths, the Rust branch has a `StorePath` type which is basically a tuple `(hash, name)` and is enforced to be correct by its constructor (so you can only pass "correct" store paths). It also doesn't store the `/nix/store` prefix. -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#3185 (comment)](https://github.com/NixOS/nix/pull/3185#issuecomment-548449035)
> > 
> > 
> > surely this would be easy to do as a cxxflag passed into the build(s)? iirc thats how its done for a bunch of stuff. that way you can determine the default location in the config options
> 
> [It is configured in `Makefile`](https://github.com/NixOS/nix/pull/3185/files#diff-a2f2a3db29e468ef4d80b60ae2ee1b4eR5-R10), not hardcoded in C++
> But that `Makefile` was made only for developing purposes, it will be replaced with meson, right?
> 
> When Nix build using Nix there is a configurable option passed down:
> 
> https://github.com/NixOS/nixpkgs/blob/fc2592b2d6348f73bfc7150f71c3df7cec0bdcd2/pkgs/tools/package-management/nix/default.nix#L2-L3

how are people going to go about building meson projects in nix? is that decided yet?

**Comment by volth** [comment]
> how are people going to go about building meson projects in nix? is that decided yet?

No, there are even more fundamental decisions have not made yet, for example:
* how to escalate privileges for symlink creation (default Windows user has no permissions to do it)
* which language to use for imperative code (currently it is perl with patched Win32::LongPath, which might be inappropriate for "people", but currently there is no other working solution)

**Comment by Ericson2314** [comment]
> How are people going to go about building meson projects in nix? is that decided yet?

@p01arst0rm I am not sure what you are asking, but we do build plenty of software using Meson in Nixpkgs today (but on Unix).

@volth my coworker proposed Lua and I must say I like that idea a lot for a new stdenv. That all said I don't think this is the right place to discus how to make Nixpkgs work on Windows. I want to emphasize that there's still tons of value we can get from Nix on Windows before that (e.g. cross compile from linux then test on Windows in separate DRV).

**Comment by conferno** [comment]
> my coworker proposed Lua and I must say I like that idea a lot for a new stdenv. That all said I don't think this is the right place to discus how to make Nixpkgs work on Windows.

While `Lua vs. other` may be offtopic here, I must explain the constraints which makes difficult using of many imperative languages originated from Unix world as `stdenv.shell`.

1. **Path is UTF-16 string** and this is difficult to avoid, because
1.1. we hit 256-byte length limit (I won't go into details now, there are some ways to workarond)
1.2. files with names which cannot be represented as UTF-8 could be on disk, after `fetchzip` or in check phase. `stdenv.shell` must be able at least `cp -r` and `rm -rf` directories which have such files inside.
Thus, internal representation of paths has to be UTF-16, so neither `bash`+`coreutils` nor stock `perl` nor stock `lua` would work out of the box :(

**Comment by Ericson2314** [comment]
I'll make a separate issue for figuring out the path types, as it will be a lot of work. Just as @volth made #2591 #2603 #2634 we can have threads about each concrete issue, and this can be the big "integration" thread.

**Comment by volth** [comment]
> @volth my coworker proposed Lua and I must say I like that idea a lot for a new stdenv. That all said I don't think this is the right place to discus how to make Nixpkgs work on Windows.

I just notice perl->python transition of `test-driver` which might be a good first step in unifying the imperative code in `nixpkgs`.
Although there are few little problems with `python` on Windows (it might need patches to support long paths and filenames with invalid UTF-8), it has a strong fanatic lobby among the contributors and replacing `bash` with `python` in `stdenv` might be a doable task in terms of volunteering workforce which would do the rewrite. It can hardly be done with Lua or something else (cc @domenkozar @FRidh)

**Comment by p01arst0rm** [comment]
> i

+1 on python for cross platform support

**Comment by p01arst0rm** [comment]
> > @volth my coworker proposed Lua and I must say I like that idea a lot for a new stdenv. That all said I don't think this is the right place to discus how to make Nixpkgs work on Windows.
> 
> I just notice perl->python transition of `test-driver` which might be a good first step in unifying the imperative code in `nixpkgs`.
> Although there are few little problems with `python` on Windows (it might need patches to support long paths and filenames with invalid UTF-8), it has a strong fanatic lobby among the contributors and replacing `bash` with `python` in `stdenv` might be a doable task in terms of volunteering workforce which would do the rewrite. It can hardly be done with Lua or something else (cc @domenkozar @FRidh)

+1 on python for cross platform support

**Comment by Ericson2314** [comment]
@volth What did you get for libbacktrace with MinGW, https://github.com/cloudwu/backtrace-mingw?

**Comment by volth** [comment]
> @volth What did you get for libbacktrace with MinGW

I do not remember.
Well, I remember the word `libbacktrace`, but `grep` founds nothing.
Where is it used in Nix?

**Comment by Ericson2314** [comment]
`BOOST_STACKTRACE_USE_BACKTRACE` in src/libutil/util.cc. I just commented that out for now.

In other news, it builds!! Will push very shortly. https://github.com/NixOS/nixpkgs/pull/73265 is the real secret sauce.

**Comment by Ericson2314** [comment]
OK build instructions

1. https://github.com/NixOS/nixpkgs/pull/73265 wait for that to hit `nixos-19.09` or switch to `release-19.09`.
2. Do a `nix-shell` for now, I haven't tested other things.
3. Add `-Ddefault_library=static` to `mesonFlags` so you do static builds, since we have no `dllimport`/`dllexport` and putting those in too soon will just cause merge hell.

Of course, I hope to automate all of this away.

**Comment by p01arst0rm** [comment]
nice work :)

**Comment by volth** [comment]
I'd suggest `with_gc` -> `with_boehm` (other `gc` too), because in Nix discourse, "GC" is mainly about garbage files in Nix Store

**Comment by Ericson2314** [comment]
Yes, that's a good idea. Also, I think I should close this one until something more up to date is ready :).

**Comment by p01arst0rm** [comment]
> Yes, that's a good idea. Also, I think I should close this one until something more up to date is ready :).

is this in regards to the meson port? are you waiting for it to be brought up to latest ?
